Update minikube context to detect outdated /etc/hosts file

nicholastmosher opened the pull request:
This is related to infinyon/fluvio#241. This encapsulates the configurations and logic needed to update the "dns context" of minikube (assigning `minikubeCA`).

A few points to note:

- This is technically a breaking change. However, in all of our repos, the only usage seemed to be in `fluvio cluster set-minikube-context`, which I am updating to match this change.
- This uses `minikube profile list -o json` in order to detect the current IP and port of minikube
- This decouples updating the kubectl context from updating the `/etc/hosts` file and minimizes shell code
  - While we're here making changes, do we want to just write Rust to edit `/etc/hosts` rather than shelling out to `sed`?

sehz commented on src/k8-config/src/context.rs:
```diff
@@ -25,71 +35,234 @@ fn load_cert_auth() -> String {
         .to_string()
 }
 
-pub struct Option {
```
this break semver requirements.  we should keep existing API.  it can be mapped new one but old code should work.

sehz commented on src/k8-config/src/context.rs:
```diff
@@ -25,71 +35,234 @@ fn load_cert_auth() -> String {
-}
 
-/// create kube context that copy current cluster configuration
-pub fn create_dns_context(option: Option) {
```
let's keep existing API for compatibility. map to new one

nicholastmosher commented:
Alright, so what I did here was I moved all of the existing code into `context::v1` and added `#[deprecated]` tags to all of the items. I added a re-export for `pub use v1::*` to the `context` module so all of those items may be imported in the same way from old clients. Then, I added the new `MinikubeContext` struct and new logic into the `context` module. I also updated the `k8-ctx-util` binary to use the new `MinikubeContext` struct.

You can test this out by checking out the `minikube` branch on the `fluvio` repo and running `cargo run --bin fluvio -- cluster set-minikube-context`. This will do the following:

- Check whether the `minikubeCA` entry in `/etc/hosts` matches the current IP address of minikube as reported by `minikube profile list -o json`.
- If the IP address is different, it will update `/etc/hosts` to match the new IP address. Otherwise, it will not touch `/etc/hosts`.
- Either way, it will update the `kubectl` context.

nicholastmosher commented:
Not sure why CI is failing, unit tests work fine on my machine. Failure seems to be related to installing nightly, I wonder if the MacOS nightly rust build is just broken. I wonder if we should mark nightly CI builds as not required for CI to pass.

sehz commented:
Similar failure happening on MacOS. Yes probably should disable nightly for now

sehz reviewed (approved):
nice work!

sehz commented:
Can you remove nightly from CI?  Rust tool chain has problem

sehz commented:
https://github.com/rust-lang/rust/issues/76638

sehz commented on src/k8-config/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "k8-config"
-version = "1.1.1"
+version = "1.1.2"
```
update to 1.2 since this is new feature
